File: web/app/policy-trace.ts
export type PolicyOrdering = 'legacy_priority' | 'scoped';
export type PolicyTrace = {
  version: string;
  ordering: PolicyOrdering;
  profiles: {
    id: string;
    name: string;
    scope: string;
    origin: 'administrator' | 'personal';
    threshold: number | null;
    selected: boolean;
  }[];
  threshold_profile: string | null;
  threshold_locked: boolean;
  rules: {
    id: string;
    name: string;
    scope: string;
    origin: 'administrator' | 'personal';
    priority: number;
    outcome: 'matched' | 'no_match' | 'missing_facts' | 'stopped';
    unavailable: string[];
    category_before: string;
    category_after: string;
    action_before: string;
    action_after: string;
    stop: boolean;
  }[];
  stopped_by: string | null;
  category_rule: string | null;
  action_rule: string | null;
  malware_override: boolean;
};
export function orderingDescription(ordering: PolicyOrdering | undefined) {
  return ordering === 'scoped'
    ? 'Personal rules run first, then administrator rules. Within each group: organization, domain, then address; ascending priority and rule ID break ties. A matching stop prevents later rules. Personal rules cannot stop administrator rules.'
    : 'Legacy ordering: only the most specific personal preference is applied. Personal rules run first, then administrator rules, in ascending priority and rule ID order.';
}
export function traceOutcome(outcome: string) {
  return (
    (
      {
        matched: 'Matched',
        no_match: 'Did not match',
        missing_facts: 'Missing facts',
        stopped: 'Skipped after stop',
      } as Record<string, string>
    )[outcome] ?? 'Unknown outcome'
  );
}
export function thresholdSource(trace: PolicyTrace) {
  if (trace.threshold_locked)
    return 'Threshold controlled by the detector decision';
  if (!trace.threshold_profile) return 'Global filter threshold';
  const profile = trace.profiles.find((p) => p.id === trace.threshold_profile);
  return profile
    ? `${profile.name} (${profile.scope})`
    : 'Recorded source unavailable';
}

export function sampleChange(row: {
  status: string;
  comparable?: boolean;
  changed?: boolean | null;
}) {
  if (row.status !== 'simulated') return 'Not available for this recipient';
  if (!row.comparable || row.changed == null) return 'Comparison incomplete';
  return row.changed ? 'Would change' : 'Unchanged';
}

File: src/policy_trace.rs
//! Recipient-scoped receipt trace; contains no rule values or message content.
use crate::{
    actions::Action,
    custom_filtering::{Field, Ordering},
    mailing::Category,
};
use serde::{Deserialize, Serialize};

pub const VERSION: &str = "recipient-policy-trace-1";
#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Origin {
    Administrator,
    Personal,
}
#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Outcome {
    Matched,
    NoMatch,
    MissingFacts,
    Stopped,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Profile {
    pub id: String,
    pub name: String,
    pub scope: String,
    pub origin: Origin,
    pub threshold: Option<f64>,
    pub selected: bool,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Rule {
    pub id: String,
    pub name: String,
    pub scope: String,
    pub origin: Origin,
    pub priority: u16,
    pub outcome: Outcome,
    pub unavailable: Vec<Field>,
    pub category_before: Category,
    pub category_after: Category,
    pub action_before: Action,
    pub action_after: Action,
    pub stop: bool,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Trace {
    pub version: String,
    pub ordering: Ordering,
    /// Most specific first; personal profile wins equal-scope ties in scoped mode.
    pub profiles: Vec<Profile>,
    pub threshold_profile: Option<String>,
    pub threshold_locked: bool,
    pub rules: Vec<Rule>,
    pub stopped_by: Option<String>,
    pub category_rule: Option<String>,
    pub action_rule: Option<String>,
    pub malware_override: bool,
}
